### PR TITLE
JSON-API: Making the tests pass

### PR DESCRIPTION
**Pull request by franzliedke**
This is still work in progress at the moment. I currently have 7 failing tests remaining.

I just wanted to get some feedback whether this is going in the right direction...

### CODE REVIEW

File: lib/roar/json/json_api.rb [hunk 1]
```diff
@@ -152,17 +153,75 @@ def to_document(res, options)
...
+          extracted["attributes"] = res unless res.empty?
+          extracted["relationships"] = relationships if relationships
+
+          extracted
+        end
+
+        def prepare_rels(rels)
+          final = {}
+          rels.each { |k, v|
+            if v.is_a?(Array)
```
Comment (apotonick) [inline lib/roar/json/json_api.rb]: Why are there so many `is_a?(Array)` here? Is JSON-API that fuzzy about its syntax? Isn't there a way to tell whether a hash entity is singular or array from the document specification (e.g. `property` vs. `collection`)?

**Comment by apotonick** [comment]
Thanks a lot! The master branch is working now, almost!

**Comment by franzliedke** [comment]
Thanks so much. Is a release planned for the next days?

**Comment by apotonick** [comment]
If you test it and tell me it's OK, then yes, otherwise no, because I don't use JSON API myself so I have zero pressure haha! :stuck_out_tongue_winking_eye:

**Comment by franzliedke** [comment]
Of course. I will report back as soon as I'm able to test it. I'm integrating this in my master's thesis, so I have the pressure. :)

**Comment by franzliedke** [comment]
Hmm, I have some problem getting it to work with `Representable::JSON::Hash`. I use hashes, because that's what I get from my backend services.

When wrapping those in a JSON-API decorator, using `Representable::JSON::Hash`, I get a result like this:

~~~json
{
  "data": {
    "type": "preference",
    "id": "",
    "attributes": {
      "id": {},
      "value": {}
    }
  }
}
~~~

Which should really look like this:
~~~json
{
  "data": {
    "type": "preference",
    "id": "some_id",
    "attributes": {
      "value": "some_value"
    }
  }
}
~~~

Once I wrap my hash in a custom `Struct` and remove the mixin, things work as expected. Any idea?

**Comment by apotonick** [comment]
I don't understand how you use those two concepts together, any sample code for me? `JSON::Hash` and `JSONAPI` are not meant to work together?!

**Comment by franzliedke** [comment]
Hmm, what's the best way then to use a JSON-API representer for data that comes as hash? Something like this:

~~~js
[
  {
    'id': 123,
    'value': 'abc'
  },
  {
    'id': 456,
    'value': 'def'
  },
  {
    'id': 789,
    'value': 'ghi'
  }
]
~~~

**Comment by apotonick** [comment]
Representers always work on top of objects that respond to public readers/writers. Array of OpenStructs, I guess?

I mean, we could think about that mechanism support by Representable out-of-the-box, but for me, that has a very low priority.

**Comment by franzliedke** [comment]
Hmm, but isn't that the whole point of `Representable::JSON::Hash`? Or did I misunderstand that?
Can you point me to what it would need to support the JSON-API stuff, too? I always get error messages about the `type` method not being defined on the hash class.

**Comment by apotonick** [comment]
No, `JSON::Hash` is meant as a standalone representer to represent hash objects. It is not designed to work with `JSONAPI` mixed in - or the other way round: `JSONAPI` is not meant to work in that environment.

Can you show me some code? Otherwise I can only guess what you're doing...

**Comment by franzliedke** [comment]
Okay, thanks for your patience.

Here's the representer module:
~~~ruby
module Preference
  include Roar::JSON::JSONAPI
  include Representable::JSON::Hash

  type :preference
  property :id
  property :value
end
~~~

It's being used like this:
~~~ruby
preferences = get_preferences_from_service # returns a hash
Preference.represent preferences
~~~

It works if I do something like this to prepare the preferences:
~~~ruby
MyStruct = Struct.new(:id, :value)
preferences.map!{|e| MyStruct.new e[:id], e[:value]}
~~~

Using something like Hashie should do the trick, too, right?

**Comment by franzliedke** [comment]
Whoops, I [just read](https://github.com/apotonick/roar/issues/171#issuecomment-159747546) that we're supposed to be using decorators with JSON-API.

So, this is what that looks like:
~~~ruby
class PreferenceDecorator < Roar::Decorator
  include Roar::JSON::JSONAPI

  type :preference
  property :id
  property :value
end
~~~

I wrap my hashes using `Hashie::Mash`:
~~~ruby
preferences.map!{ |e| Hashie::Mash.new e }
~~~

Now, rendering works, but only for one of them:
~~~ruby
PreferenceDecorator.prepare preference.first
~~~

But not for an array:
~~~ruby
PreferenceDecorator.for_collection.prepare preferences
~~~

...because `collection` is `nil` on [this line](https://github.com/apotonick/roar/blob/3540c8ea27cbd587f711ea412c264f6d7b02e355/lib/roar/json/json_api.rb#L40).

**Comment by apotonick** [comment]
Don't use Hashie! Use `OpenStruct`.

```ruby
prefs = preferences.collect { |p| OpenStruct.new(p) }
PreferenceDecorator.new(prefs).to/from_json ...
```

**Pull request closed without merging**